File: ruby/all-your-base/all_your_base.rb
class BaseConverter
  def self.convert(in_base, digits, out_base)
    raise ArgumentError unless valid_digits?(digits, in_base, out_base)

    base10 = to_base_10(digits, in_base)
    from_base_10(base10, out_base)
  end

  def self.valid_digits?(digits, in_base, out_base)
    in_base > 1 && out_base > 1 && digits.all? { |n| (0...in_base).cover?(n) }
  end

  def self.to_base_10(digits, base)
    digits.reverse.each_with_index.reduce(0) do |sum, number_and_index|
      number, index = number_and_index
      sum + (number * base**index)
    end
  end

  def self.from_base_10(number, base)
    n = number
    digits = []

    while n.positive?
      digits << n.remainder(base)
      n = (n / base).floor
    end

    digits.empty? ? [0] : digits.reverse
  end
end

File: clojure/all-your-base/src/all_your_base.clj
(ns all-your-base)

(defn- pow [a b]
  (reduce * 1 (repeat b a)))

(defn- floor
  ([a] (floor a 1))
  ([a divisor] (int (Math/floor (/ a divisor)))))

(defn- to-base-10 [digits base]
  (->> (reverse digits)
       (map-indexed #(* %2 (pow base %1)))
       (apply +)))

(defn- from-base-10 [number base]
  (loop [digits '()
         n number]
    (println)
    (if (pos? n)
      (recur (conj digits (rem n base))
             (floor n base))
      (if (empty? digits)
        '(0)
        digits))))

(defn- valid-base-conversion? [digits from-base to-base]
  (and (> from-base 1)
       (> to-base 1)
       (pos? (count digits))
       (every? #(> from-base % -1) digits)))

(defn convert [from-base digits to-base]
  (if (valid-base-conversion? digits from-base to-base)
    (-> digits
        (to-base-10 from-base)
        (from-base-10 to-base))))
